### ISSUE TITLE
Many releases do not build

### ISSUE THREAD

**Issue by andrew-wja** [oneapi-src/oneDNN#707]
# Summary
onednn/dnnl/mkl-dnn versions 0.X.Y are used by many projects, including the Intel Caffe distribution. However, the vast majority of these tagged releases are broken, and do not build. It appears that someone enabled warnings-as-errors by default long ago, and since every release I have tried on the 0.17, 0.18, 0.19 and 0.20 branches has at least one warning during build, there are no releases for those versions which actually can be built!

# Version
v0.19 41bee20d7eb4a67feeeeb8d597b3598994eb1959
v0.18.1 7de7e5d02bf687f971e7668963649728356e0c20
v0.18 863ff6e7042cec7d2e29897fe9f0872e0888b0fc
v0.17.{4,3,2,1} 
etc.

# Environment
n/a

# Steps to reproduce
Just download any of the relevant release archives and run the build script with default options

# Observed behavior
None of these releases build because they have warnings, and `-Werror` is turned on. For some of them, there is just one single trivial warning that is causing the build to be broken!

# Expected behavior
Released archives should build if the default build options are used. Otherwise, they are not functional releases.

Is it possible to fix this issue by disabling `-Werror` in release archives? There is no point in having "releases" which do not build due to a misconfiguration of the build system. Many of these releases work just fine if `-Werror` is disabled, but that requires the build systems of many downstream pieces of software to be changed.

**Comment by andrew-wja** [comment]
Just an example, trying to build https://github.com/intel/mkl-dnn/releases/download/v0.17-rc/mklml_lnx_2019.0.1.20180928.tgz

```
In file included from /usr/include/string.h:495,
                 from /home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:17:
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘int mkldnn::impl::mkldnn_getenv(char*, const char*, int)’ at /home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:49:24:
/usr/include/bits/string_fortified.h:106:34: error: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp: In function ‘int mkldnn::impl::mkldnn_getenv(char*, const char*, int)’:
/home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:45:34: note: length computed here
   45 |             value_length = strlen(buffer);
      |                            ~~~~~~^~~~~~~~
In file included from /usr/include/string.h:495,
                 from /home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:17:
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘int mkldnn::impl::mkldnn_getenv(char*, const char*, int)’ at /home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:49:24,
    inlined from ‘bool mkldnn::impl::mkldnn_jit_dump()’ at /home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:68:26:
/usr/include/bits/string_fortified.h:106:34: error: ‘char* __builtin___strncpy_chk(char*, const char*, long unsigned int, long unsigned int)’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp: In function ‘bool mkldnn::impl::mkldnn_jit_dump()’:
/home/andrew/.cache/yay/intel-caffe-git/src/intel-caffe/external/mkldnn/src/src/common/utils.cpp:45:34: note: length computed here
   45 |             value_length = strlen(buffer);
      |                            ~~~~~~^~~~~~~~
cc1plus: all warnings being treated as errors
```

This same behaviour is observed for all 0.17.X, 0.18.X, 0.19.X release archives, just with slightly different warnings.

**Comment by rsdubtso** [comment]
Thanks for the report. It is unfortunate that the package does not build. I can retroactively disable the `-Werror` in the old branches, but I cannot do this for releases because they are tied to a hash and re-tagging releases does not seem like a particularly good idea since it also may break things. 

Additionally, IntelCaffe uses a particular dnnl commit (https://github.com/intel/caffe/blob/master/mkldnn.commit) and thus won't be affected at all. Hence, package mainteiners (AUR?) need to work around this.

**Comment by andrew-wja** [comment]
Thanks for the reply! Sure, it is straightforward to work around the issue on the maintainer side, for example, I can add a line to the package build that replaces the contents of mkldnn.commit with `rls-v0.19`, which is where that hash should be pointing (development on Intel Caffe seems to have stopped).

However, the bigger issue is that if I `git checkout rls-v0.19` in oneDNN, it doesn't build!

Modifying the release tarballs would indeed be a bit dangerous. Would it be possible to add a commit to the top of the `rls-vY.XX` branches disabling -Werror though?

**Comment by rsdubtso** [comment]
> However, the bigger issue is that if I git checkout rls-v0.19 in oneDNN, it doesn't build!
I need more info. Works for me with gcc-9.1 and gcc-9.3.

> Would it be possible to add a commit to the top of the rls-vY.XX branches disabling -Werror though?
Yes, I'll 'fix' everything dating back to 0.17. The warning above is actually a bug that was fixed in rls-v0.19, so you already can go ahead and fix the hash in intel-caffe to point to 027de7603662a569366e15132ac80298902b96b8. The fix just have not been backported far enough...

**Comment by andrew-wja** [comment]
> > However, the bigger issue is that if I git checkout rls-v0.19 in oneDNN, it doesn't build!
> 
> I need more info. Works for me with gcc-9.1 and gcc-9.3.
> 
> > Would it be possible to add a commit to the top of the rls-vY.XX branches disabling -Werror though?
> 
> Yes, I'll 'fix' everything dating back to 0.17. The warning above is actually a bug that was fixed in rls-v0.19, so you already can go ahead and fix the hash in intel-caffe to point to [027de76](https://github.com/oneapi-src/oneDNN/commit/027de7603662a569366e15132ac80298902b96b8). The fix just have not been backported far enough...

Apologies, I mistyped: the rls-v0.19 branch specifically *does* build correctly. I have managed to get Intel Caffe to build and run again against rls-v0.19 with only minor changes (see https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=intel-caffe-git)

Since rls-v0.19 has the fix, the issue is fixed-for-me at this point, so feel free to close this issue when you need to!

**Comment by rsdubtso** [comment]
Fixed
